rfcs/2021-11-05-9930-native-event-encoding.md

```diff
@@ -0,0 +1,186 @@
+- Both `LogEvent` and `Metric` derive a `serde::Deserialize` implementation already, while `Metric`
+  also derives a `serde::Serialize` implementation.
+- We would add missing `serde` derives to `LogEvent` and `Event` itself, allowing top-level `Event`s
+  to be trivially serialized and deserialized.
+- We would _not_ serialize/deserialize event metadata, which currently only includes event
+  finalizers and a Datadog API key override field.
+- Event metadata may come into scope in the future where there is a more generalized mechanism for
+  adding metadata to events, but we would need to design a mechanism to filter "internal" metadata
+  vs "external" metadata, as we would not want to push API keys in plaintext, etc.
+- The existing framing/codec work happening for both sources and sinks would gain two new
+  implementations for `vector_native` and `vector_json`, respectively.
+- We would use `serde-reflection` to generate a basic schema of `Event`, which could be stored in
+  the source code itself (similar in principle to Cargo.lock).  This would serve as the minimum
+  viable schema for JSON use cases, whichout any commitment to versioning or backwards/forwards-compatibility.
+
+## Rationale
+
+Adding encodings for natively representing events would provide an additional avenue for users to
+both ingest data into Vector, as well as constructing more complex Vector deployment topologies.  As
+Vector development can often be bottlenecked when it comes to add new sources and sink, this work
+would act as a force multiplier for letting users invest a small amount of time converting their
+data to the native format, and then being able to universally ingest it.
+
+If we didn't do this, it would not necessarily hurt the long-term goals of Vector, but it would
+require more effort over time in order to develop new sources and sinks to meet the demands of users
+who wish to use Vector with systems we don't already support.  This could hurt the long-term
+_success_ of Vector.
+
+## Drawbacks
+
+- Why should we not do this?
+- What kind on ongoing burden does this place on the team?
+
+Encoding `Event` natively via Protocol Buffers should be a feature we can accomplish with no
+additional burden on the Vector team, as we already perform the necessary due diligence and spend
+time ensuring that our Protocol Buffers definition stays backwards/forwards-compatible.
+
+Encoding `Event` to JSON, however, could result in more time spent by the Vector team on support to
+the lack of a stringent schema, given that we would not be transforming `Event` to a known
+definition like Protocol Buffers.  While the idea to generate a minimal viable schema could
+hopefully alleviate some of those concerns, it still doesn't address the notion of not providing
+versioned JSON schemas or backwards/forwards-compatibility, which would likely represent the bulk of
+issues brought to us for users utilizing `vector_json`.
```

Comment:
One additional benefit of this being a codec is that it would give us one centralized location to do this kind of maintenance, add compatibility layers, etc. If we were relying directly on `serde` in each source/sink, we'd need to do all kinds of serde magic on the event representations themselves to handle those kinds of things. It seems like with this approach, we could theoretically move away from serde entirely sometime down the line if we decide it's worthwhile to do something more manual to either decouple or better handle compatibility.